Add TiDB Cloud Terraform Doc (#10563)
pingcap/docs: tidb-cloud/terraform-use-cluster-resource.md

---

> title: Use Cluster Resource

Reviewer: ```suggestion
title: Use Cluster Resources
```

> summary: Learn how to use the cluster resource to create and modify a TiDB Cloud cluster.

Reviewer: ```suggestion
summary: Learn how to use the cluster resources to create and modify a TiDB Cloud cluster.
```

Author: cluster resource 是一个连贯的专有名词

---

> # Use Cluster Resource

Reviewer: ```suggestion
# Use Cluster Resources
```

Author: not Resources. cluster resource 是一个连贯的专有名词

You can learn how to create and modify a TiDB Cloud cluster with the `tidbcloud_cluster` resource in this document.

> Besides, you will also learn how to get the necessary information with the `tidbcloud_project` and `tidbcloud_cluster_spec` data sources.

Reviewer: ```suggestion
In addition, you will also learn how to get the necessary information with the `tidbcloud_project` and `tidbcloud_cluster_spec` data sources.
```

                      + node_quantity    = 1
                      + node_size        = "8C64G"
                      + storage_size_gib = 500
                    }
                    # (2 unchanged attributes hidden)
                }
                # (3 unchanged attributes hidden)
            }
            id             = "1379661944630234067"
            name           = "firstCluster"
          ~ status         = "AVAILABLE" -> (known after apply)
            # (4 unchanged attributes hidden)
        }

    Plan: 0 to add, 1 to change, 0 to destroy.

    Do you want to perform these actions?
      Terraform will perform the actions described above.
      Only 'yes' will be accepted to approve.

      Enter a value:

    ```

    As in the above execution plan, TiFlash will be added, and one resource will be changed.

3. If everything in your plan looks fine, type `yes` to continue:

    ```
      Enter a value: yes

    tidbcloud_cluster.example_cluster: Modifying... [id=1379661944630234067]
    tidbcloud_cluster.example_cluster: Modifications complete after 2s [id=1379661944630234067]

    Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
    ```

> 4. Use `terraform state show tidbcloud_cluster.example_cluster` to see the status:

Reviewer 2: ```suggestion
4. Use `terraform state show tidbcloud_cluster.example_cluster` to see the status. Note that you need to replace `example_cluster` with your cluster resource name.
```

Author: you need to -> you can

Author: 感觉在最开始创建的 cluster.tf 的时候说一下可以替换就可以了。后面的操作不必每次都提

                }
                tiflash = {
                    node_quantity    = 1
                    node_size        = "8C64G"
                    storage_size_gib = 500
                }
                tikv    = {
                    node_quantity    = 3
                    node_size        = "8C32G"
                    storage_size_gib = 500
                }
            }
            ip_access_list = [
                # (1 unchanged element hidden)
            ]
            port           = 4000
            root_password  = "Your_root_password1."
        }
        id             = "1379661944630234067"
        name           = "firstCluster"
        project_id     = "1372813089189561287"
        region         = "eu-central-1"
        status         = "MODIFYING"
    }
    ```

The `MODIFYING` status indicates that the cluster is changing now. Wait for a moment. The status will be changed to `AVAILABLE`.

### Scale a TiDB cluster

You can scale a TiDB cluster when its status is `AVAILABLE`.

1. In the `cluster.tf` file that is used when you [create the cluster](#create-a-cluster-using-the-cluster-resource), edit the `components` configurations.

>     For example, to add one more node for TiDB, 3 more nodes for TiKV (TiKV needs to add at least 3 nodes for its step is 3. You can [get this information from the cluster specifcation](#get-cluster-specification-information-using-the-cluster-spec-data-source)), and one more node for TiFlash, you can edit the configurations as follows:

Reviewer: "for its step is 3" looks a bit confusing.
```suggestion
    For example, to add one more node for TiDB, 3 more nodes for TiKV (TiKV needs at least 3 nodes for its step is 3. You can [get this information from the cluster specifcation](#get-cluster-specification-information-using-the-cluster-spec-data-source)), and one more node for TiFlash, you can edit the configurations as follows:
```

Author: 这里删除 to add 意思会变。这里表达的是步长是3，扩容时需要是 3的倍数。我优化一些这一段

>     For example, to add one more node for TiDB, 3 more nodes for TiKV (TiKV nodes needs to be a multiple of 3 for its step is 3. You can [get this information from the cluster specifcation](#get-cluster-specification-information-using-the-cluster-spec-data-source)), and one more node for TiFlash, you can edit the configurations as follows:

Reviewer 2: ```suggestion
    For example, to add one more node for TiDB, 3 more nodes for TiKV (The number of TiKV nodes needs to be a multiple of 3 for its step is 3. You can [get this information from the cluster specifcation](#get-cluster-specification-information-using-the-cluster-spec-data-source)), and one more node for TiFlash, you can edit the configurations as follows:
```

                     ~ node_quantity = 1 -> 2
                       # (1 unchanged attribute hidden)
                   }
                 ~ tiflash = {
                     ~ node_quantity    = 1 -> 2
                       # (2 unchanged attributes hidden)
                   }
                 ~ tikv    = {
                     ~ node_quantity    = 3 -> 6
                       # (2 unchanged attributes hidden)
                   }
               }
               # (3 unchanged attributes hidden)
           }
           id             = "1379661944630234067"
           name           = "firstCluster"
         ~ status         = "AVAILABLE" -> (known after apply)
           # (4 unchanged attributes hidden)
       }

   Plan: 0 to add, 1 to change, 0 to destroy.

   Do you want to perform these actions?
     Terraform will perform the actions described above.
     Only 'yes' will be accepted to approve.

     Enter a value: yes

   tidbcloud_cluster.example_cluster: Modifying... [id=1379661944630234067]
   tidbcloud_cluster.example_cluster: Modifications complete after 2s [id=1379661944630234067]

   Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
   ```

> Wait for the status from `MODIFYING` to `AVAILABLE`.

Reviewer: ```suggestion
Wait for the status to turn from `MODIFYING` to `AVAILABLE`.
```

Author: 这样不通顺？

   Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
     ~ update in-place

   Terraform will perform the following actions:

     # tidbcloud_cluster.example_cluster will be updated in-place
     ~ resource "tidbcloud_cluster" "example_cluster" {
         ~ config         = {
             + paused         = true
               # (4 unchanged attributes hidden)
           }
           id             = "1379661944630234067"
           name           = "firstCluster"
         ~ status         = "AVAILABLE" -> (known after apply)
           # (4 unchanged attributes hidden)
       }

   Plan: 0 to add, 1 to change, 0 to destroy.

   Do you want to perform these actions?
     Terraform will perform the actions described above.
     Only 'yes' will be accepted to approve.

     Enter a value: yes

   tidbcloud_cluster.example_cluster: Modifying... [id=1379661944630234067]
   tidbcloud_cluster.example_cluster: Modifications complete after 2s [id=1379661944630234067]

   Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
   ```

> 3. Use the `terraform state show tidbcloud_cluster.example_cluster` command to check the status:

Reviewer 2: ```suggestion
3. Use the `terraform state show tidbcloud_cluster.example_cluster` command to check the status. Note that you need to replace `example_cluster` with your cluster resource name.
```

                   node_size        = "8C64G"
                   storage_size_gib = 500
               }
               tikv    = {
                   node_quantity    = 6
                   node_size        = "8C32G"
                   storage_size_gib = 500
               }
           }
           ip_access_list = [
               # (1 unchanged element hidden)
           ]
           paused         = false
           port           = 4000
           root_password  = "Your_root_password1."
       }
       id             = "1379661944630234067"
       name           = "firstCluster"
       project_id     = "1372813089189561287"
       region         = "eu-central-1"
       status         = "RESUMING"
   }
   ```

6. Wait for a moment, then use the `terraform refersh` command to update the state. The status will be changed to `AVAILABLE` finally.

Now, you have created and managed a Dedicated Tier cluster with Terraform. Next, you can try creating a backup of the cluster by our [backup resource](/tidb-cloud/terraform-use-backup-resource.md).

## Import a cluster

You can manage a cluster with Terraform by importing the cluster even if it is not created by Terraform.

> 1. Create a new `import_cluster.tf` as follows:

Reviewer 2: ```suggestion
1. Create a `import_cluster.tf` file as follows:
```

---
title: Use Cluster Resource
summary: Learn how to use the cluster resource to create and modify a TiDB Cloud cluster.
---

# Use Cluster Resource

You can learn how to create and modify a TiDB Cloud cluster with the `tidbcloud_cluster` resource in this document.

In addition, you will also learn how to get the necessary information with the `tidbcloud_project` and `tidbcloud_cluster_spec` data sources.

## Prerequisites

- [Get TiDB Cloud Terraform Provider](/tidb-cloud/terraform-get-tidbcloud-provider.md).

## Get project IDs using the project data source

> Each TiDB cluster is in a project. Before you create a TiDB cluster, you need to get the ID of the project that you want to create a cluster in.

Reviewer: ```suggestion
Each TiDB cluster is in a project. Before you create a TiDB cluster, you need to get the ID of the project in which you want to create a cluster.
```

               }
               tiflash = {
                   node_quantity    = 2
                   node_size        = "8C64G"
                   storage_size_gib = 500
               }
               tikv    = {
                   node_quantity    = 6
                   node_size        = "8C32G"
                   storage_size_gib = 500
               }
           }
           ip_access_list = [
               # (1 unchanged element hidden)
           ]
           paused         = false
           port           = 4000
           root_password  = "Your_root_password1."
       }
       id             = "1379661944630234067"
       name           = "firstCluster"
       project_id     = "1372813089189561287"
       region         = "eu-central-1"
       status         = "RESUMING"
   }
   ```

6. Wait for a moment, then use the `terraform refersh` command to update the state. The status will be changed to `AVAILABLE` finally.

Now, you have created and managed a Dedicated Tier cluster with Terraform. Next, you can try creating a backup of the cluster by our [backup resource](/tidb-cloud/terraform-use-backup-resource.md).

## Import a cluster

> You can manage a cluster with Terraform by importing the cluster even if it is not created by Terraform.

Reviewer 2: ```suggestion
You can import a cluster in the following scenarios:

- Use Terraform to manage a cluster that is not created by Terraform. 
- Use Terraform to restore a cluster that is [created by a restore task with the restore resource](/tidb-cloud/terraform-use-restore-resource.md#create-a-restore-task-with-the-restore-resource). 
```

> The following steps show you how to import the cluster created by the restore task in the last section.

Reviewer 2: Do we need this sentence?

> 1. Create a new `import_cluster.tf` as follows:

Reviewer 2: Do we need to create a directory first?

Author: both can

Author: 我是觉得一个小节建一个 directory 避免当前目录影响我们的教程就可以了，所以这边没让新建

    ```
    terraform {
     required_providers {
       tidbcloud = {
         source = "tidbcloud/tidbcloud"
         version = "~> 0.0.1"
       }
     }
     required_version = ">= 1.0.0"
   }
    resource "tidbcloud_cluster" "import_cluster" {}
    ```

2. Import the cluster by `terraform import tidbcloud_cluster.import_cluster projectId,clusterId`:

   For example:

    ```
    $ terraform import tidbcloud_cluster.import_cluster 1372813089189561287,1379661944630264072

    tidbcloud_cluster.import_cluster: Importing from ID "1372813089189561287,1379661944630264072"...
    tidbcloud_cluster.import_cluster: Import prepared!
      Prepared tidbcloud_cluster for import
    tidbcloud_cluster.import_cluster: Refreshing state... [id=1379661944630264072]

    Import successful!

    The resources that were imported are shown above. These resources are now in
    your Terraform state and will henceforth be managed by Terraform.
    ```

3. Run the `terraform state show tidbcloud_cluster.import_cluster` command to check the status of the cluster:

    ```

>     $ terraform state show tidbcloud_cluster.restore_cluster1

Reviewer 2: 这里的 restore_cluster1 是否要改成 import_cluster

Author: 是的 改漏了

    # tidbcloud_cluster.restore_cluster1:

>     resource "tidbcloud_cluster" "restore_cluster1" {

Reviewer 2: 同上

        cloud_provider = "AWS"
        cluster_type   = "DEDICATED"
        config         = {
            components = {
                tidb    = {
                    node_quantity = 2
                    node_size     = "8C16G"
                }
                tiflash = {
                    node_quantity    = 2
                    node_size        = "8C64G"
                    storage_size_gib = 500
                }
                tikv    = {
                    node_quantity    = 6
                    node_size        = "8C32G"
                    storage_size_gib = 500
                }
            }
            port       = 4000
        }
        id             = "1379661944630264072"
        name           = "restoreCluster"
        project_id     = "1372813089189561287"
        region         = "eu-central-1"
        status         = "AVAILABLE"
    }
    ```

4. To manage the cluster using Terraform, you can copy the output of the previous step to your configuration file. Note that you need to delete the lines of `id` and `status`, because they will be controlled by Terraform instead:

    ```

>     resource "tidbcloud_cluster" "restore_cluster1" {

Reviewer 2: 同上

                      storage_size_gib = 500
                  }
                  tikv    = {
                      node_quantity    = 6
                      node_size        = "8C32G"
                      storage_size_gib = 500
                  }
              }
              port       = 4000
          }
          name           = "restoreCluster"
          project_id     = "1372813089189561287"
          region         = "eu-central-1"
    }
    ```

5. You can use `terraform fmt` to format your configuration file:

    ```
    $ terraform fmt
    ```

6. To ensure the consistency of the configuration and state, you can execute `terraform plan` or `terraform apply`. If you see `No changes`, the import is successful.

    ```
    $ terraform apply

    tidbcloud_cluster.import_cluster: Refreshing state... [id=1379661944630264072]

    No changes. Your infrastructure matches the configuration.

    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

    Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
    ```

Now you can use Terraform to manage the cluster. 

> 
> An example of using import is managing the cluster created by restored task.

Reviewer 2: ```suggestion
```
